### PR TITLE
Add kmem cache accessors

### PR DESCRIPTION
**Pull request by mattmacy**
Make metaslab.c platform agnostic again

Signed-off-by: Matt Macy <mmacy@FreeBSD.org>

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by ikozhukhov** [comment]
they are changes are not dilos friendly and need prepare additional wrapper.
probably in this time, for reduce collisions, will be much more better to rename functions with prefix `zfs_kmem_*` and prepare platform specific updates

**Comment by mattmacy** [comment]
> they are changes are not dilos friendly and need prepare additional wrapper.
> probably in this time, for reduce collisions, will be much more better to rename functions with prefix `zfs_kmem_*` and prepare platform specific updates

If you want something specifically that works on Dilos, you'll need to fork it and submit a PR to my branch. Do these specific functions conflict with existing kmem functions on Dilos? I know that Illumos has some super heavy weight string based KPI for accessing properties like this.

**Comment by behlendorf** [comment]
@ ikozhukhov we've definitely prefixed other functions where the was a collision, but do `spl_kmem_cache_inuse` and `spl_kmem_entry_size` conflict with existing functions?

**Comment by freqlabs** [comment]
I don't see anything with these names in illumos-gate:
http://src.illumos.org/source/search?q=kmem_cache_inuse&project=illumos-gate
http://src.illumos.org/source/search?q=kmem_cache_entry_size&project=illumos-gate

**Comment by ikozhukhov** [comment]
yes, they are not defined on illumos, but can be implemented in different manner. for reduce collisions in future, can be move name `kmem_cache_inuse` to `zfs_kmem_cache_inuse` and implement wrapper per platform?

**Comment by behlendorf** [comment]
@ikozhukhov for the moment I'd prefer to keep this change small.  But I agree there's a bigger discussion we may want to have about adding a `zfs_` prefix to all of the `kmem_cache_*` functions.  At least one of these functions already conflicts with a Linux symbol which requires some ugly workarounds.

File: include/os/linux/spl/sys/kmem_cache.h [hunk 1]
```diff
@@ -218,6 +218,8 @@ extern void spl_kmem_cache_free(spl_kmem_cache_t *skc, void *obj);
 extern void spl_kmem_cache_set_allocflags(spl_kmem_cache_t *skc, gfp_t flags);
 extern void spl_kmem_cache_reap_now(spl_kmem_cache_t *skc, int count);
 extern void spl_kmem_reap(void);
+extern int spl_kmem_cache_inuse(kmem_cache_t *cache);
+extern int spl_kmem_cache_entry_size(kmem_cache_t *cache);
```
[the remark this answers is not in the document]
Comment (mattmacy) [reply include/os/linux/spl/sys/kmem_cache.h]: good point.

**Comment by mattmacy** [comment]
> yes, they are not defined on illumos, but can be implemented in different manner. for reduce collisions in future, can be move name `kmem_cache_inuse` to `zfs_kmem_cache_inuse` and implement wrapper per platform?

@ikozhukhov I'd like to prefix common ZFS code with `zfs_` and SPL interfaces with `spl_`

**Comment by behlendorf** [comment]
@lundman would you mind reviewing this.

**Comment by lundman** [comment]
The change itself look fine, I guess it comes down to what `cache->skc_obj_total` and `cache->skc_obj_size` refer to, to find equivalent data. Does `skc_obj_size` mean the size used when creating range_seg_cache? As in:
```
range_seg_cache = kmem_cache_create("range_seg_cache",
	    sizeof (range_seg_t), 0, NULL, NULL, NULL, NULL, NULL, 0);
```
Ie, the static `sizeof(range_seg_t)` but added as a method? (Which is a good thing, just making sure I follow) Which would make it `cache->cache_bufsize` in illumos.

And `skc_obj_total` is added-up total, so we know to evict. kstat appears to calculate it:
```
	for (cpu_seqid = 0; cpu_seqid < max_ncpus; cpu_seqid++) {
		kmem_cpu_cache_t *ccp = &cp->cache_cpu[cpu_seqid];
		cpu_buf_avail = 0;
			cpu_buf_avail += ccp->cc_rounds;
			cpu_buf_avail += ccp->cc_prounds;
		buf_avail			+= cpu_buf_avail;
	}
	buf_avail += cp->cache_full.ml_total * cp->cache_magtype->mt_magsize;
	buf_avail += cp->cache_bufslab;
	kmcp->kmc_buf_inuse.value.ui64	= cp->cache_buftotal - buf_avail;
```
Ok, I think I know what to do - looks good to me :)

**Comment by behlendorf** [comment]
@mattmacy looks like this needs a rebase to resolve a conflict with a recent merge.  Then this is ready to merge.

**Comment by behlendorf** [comment]
Almost there, there's a cstyle issue after the rebase.

```
./module/zfs/metaslab.c: 2183: line > 80 characters
Makefile:1613: recipe for target 'cstyle' failed
```

**Pull request merged**